extensions/data-plane/data-plane-spi/src/main/java/org/eclipse/dataspaceconnector/dataplane/spi/pipeline/ParallelSink.java

```diff
@@ -0,0 +1,102 @@
+/*
+ *  Copyright (c) 2021 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+package org.eclipse.dataspaceconnector.dataplane.spi.pipeline;
+
+import org.eclipse.dataspaceconnector.common.stream.PartitionIterator;
+import org.eclipse.dataspaceconnector.dataplane.spi.result.TransferResult;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.AbstractResult;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.dataspaceconnector.common.async.AsyncUtils.asyncAllOf;
+import static org.eclipse.dataspaceconnector.spi.response.ResponseStatus.ERROR_RETRY;
+
+/**
+ * Writes data in parallel.
+ */
+public abstract class ParallelSink implements DataSink {
+    protected String requestId;
+    protected int partitionSize = 5;
+    protected ExecutorService executorService;
+    protected Monitor monitor;
+
+    @Override
+    public CompletableFuture<TransferResult> transfer(DataSource source) {
```

Comment:
Should we unit test this method in isolation from the implementing classes?

Reply:
Created https://github.com/agera-edc/DataSpaceConnector/issues/99 (since this is existing code and orthogonal to implementation of Azure storage)